src/Aquifer.Common/Services/Caching/CachingLanguageService.cs

```diff
@@ -57,11 +70,113 @@ public async Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(
                 cacheEntry.SlidingExpiration = s_cacheLifetime;
 
                 var LanguageCodeById = await _dbContext.Languages
-                    .ToDictionaryAsync(li => li.Id, li => li.ISO6393Code, ct);
+                    .Select(l => new { l.Id, l.ISO6393Code })
+                    .ToDictionaryAsync(l => l.Id, l => l.ISO6393Code, ct);
 
                 return
                     (LanguageCodeById,
-                    LanguageIdByCode: LanguageCodeById.ToDictionary(li => li.Value, li => li.Key));
+                    LanguageIdByCode: LanguageCodeById.ToDictionary(li => li.Value, li => li.Key, StringComparer.OrdinalIgnoreCase));
             });
     }
-}
+
+    public async Task<(bool IsValidLanguageId, bool IsValidLanguageCode, int? ValidLanguageId)>
+        ValidateLanguageIdOrCodeAsync(int? languageId, string? languageCode, bool shouldRequireInput, CancellationToken ct)
+    {
+        if (languageId is not null && languageCode is not null)
+        {
+            throw new ArgumentException($"Only one of {nameof(languageId)} or {nameof(languageCode)} may be provided.");
+        }
+
+        if (languageId is not null)
+        {
+            return (
+                IsValidLanguageId: await GetLanguageCodeAsync(languageId.Value, ct) is not null,
+                IsValidLanguageCode: true,
+                languageId.Value);
+        }
+
+        if (languageCode is not null)
+        {
+            var languageIdForCode = await GetLanguageIdAsync(languageCode, ct);
+            return (
+                IsValidLanguageId: true,
+                IsValidLanguageCode: languageIdForCode is not null,
+                languageIdForCode.GetValueOrDefault());
+        }
+
+        if (!shouldRequireInput)
+        {
+            return (
+                IsValidLanguageId: true,
+                IsValidLanguageCode: true,
+                ValidLanguageId: null);
+        }
+
+        throw new ArgumentException($"One of {nameof(languageId)} or {nameof(languageCode)} must be non-null.");
+    }
+
+    public async Task<
+            (IReadOnlyList<int> InvalidLanguageIds,
+            IReadOnlyList<string> InvalidLanguageCodes,
+            IReadOnlyList<int>? ValidLanguageIds)>
+        ValidateLanguageIdsOrCodesAsync(
```

Comment:
I moved a couple of validation methods here so they can easily be used by all of the endpoints that deal with language IDs or Codes.